### PR TITLE
Add Web Benchmarks

### PR DESCRIPTION
**Pull request by pennzht**
This PR adds web benchmarks to the gallery.

### CODE REVIEW

File: lib/benchmarks/gallery_recorder.dart [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    this.shouldRunPredicate,
+    this.testScrollsOnly = false,
+  }) : assert(testScrollsOnly || shouldRunPredicate != null),
+        super(name: benchmarkName, useCustomWarmUp: true);
+
+  /// The name of the gallery benchmark to be run.
+  final String benchmarkName;
+
+  /// A function that accepts the name of a demo and returns whether we should
+  /// run this demo in this benchmark.
```
Comment (guidezpl) [inline lib/benchmarks/gallery_recorder.dart:24]: Refer to a file or class in this comment to help determine in which format a demo name should be written

File: lib/benchmarks/gallery_recorder.dart [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  final bool Function(String) shouldRunPredicate;
+
+  /// Whether this benchmark only tests scrolling.
+  final bool testScrollsOnly;
+
+  /// Whether we should continue recording.
+  @override
```
Comment (guidezpl) [inline lib/benchmarks/gallery_recorder.dart:31]: nit: move after _finished to maintain a logical class order

File: pubspec.yaml [hunk 3]
```diff
@@ -32,6 +32,8 @@ dependencies:
   provider: ^4.3.2
   adaptive_breakpoints: ^0.0.2
 
+  web_benchmarks: ^0.0.2
```
Comment (guidezpl) [inline pubspec.yaml:35]: ⇒ dev_dependencies

File: test/benchmarks_test.dart [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import 'dart:convert' show JsonEncoder;
+import 'dart:io';
+
+import 'package:test/test.dart';
+
+import 'package:gallery/benchmarks/common.dart';
+
+import 'package:web_benchmarks/server.dart';
+
+final listOfBenchmarks = <String>[
```
Comment (guidezpl) [inline test/benchmarks_test.dart:10]: testing a benchmark test... nice 👍

File: test/benchmarks_test.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  galleryUnanimatedPerf,
+  galleryAnimatedPerf,
+  galleryScrollPerf,
+];
+
+Future<void> main() async {
+  test('Can run a web benchmark', () async {
+    print ('Starting web benchmark tests ...');
+
+    final taskResult = await serveWebBenchmark(
```
Comment (guidezpl) [inline test/benchmarks_test.dart:21]: Might be better to run `run_benchmarks.dart`'s `main` directly instead

File: test/benchmarks_test.dart [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import 'dart:convert' show JsonEncoder;
+import 'dart:io';
+
+import 'package:test/test.dart';
```
Comment (guidezpl) [inline test/benchmarks_test.dart:4]: nit: reorder

File: test/benchmarks_test.dart (hunk 4 above)
Comment (guidezpl) [inline test/benchmarks_test.dart:10]: consider moving this to common.dart

File: lib/benchmarks/gallery_recorder.dart [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+/// A recorder that measures frame building durations for the Gallery.
+class GalleryRecorder extends WidgetRecorder {
+  GalleryRecorder({
+    @required this.benchmarkName,
+    this.shouldRunPredicate,
+    this.testScrollsOnly = false,
+  }) : assert(testScrollsOnly || shouldRunPredicate != null),
+        super(name: benchmarkName, useCustomWarmUp: true);
+
+  /// The name of the gallery benchmark to be run.
```
Comment (guidezpl) [inline lib/benchmarks/gallery_recorder.dart:20]: mention that names are available in `common.dart`

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

File: lib/benchmarks/gallery_recorder.dart [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  final String benchmarkName;
+
+  /// A function that accepts the name of a demo and returns whether we should
+  /// run this demo in this benchmark.
+  final bool Function(String) shouldRunPredicate;
+
+  /// Whether this benchmark only tests scrolling.
+  final bool testScrollsOnly;
```
[the remark this answers is not in the document]
Comment (guidezpl) [reply lib/benchmarks/gallery_recorder.dart:28]: can you add more info about what this does?

**Comment by guidezpl** [comment]
Are there plans to add this to CI in a later PR?

File: lib/benchmarks/gallery_recorder.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (pennzht) [reply lib/benchmarks/gallery_recorder.dart:28]: When this is set to `true`, the automator will only do scrolling on the page (will not open studies or demos).

Maybe `testScrollingOnly`.

File: lib/benchmarks/gallery_recorder.dart (hunk 1 above)
Comment (pennzht) [reply lib/benchmarks/gallery_recorder.dart:24]: Added some more documentation.

File: lib/benchmarks/gallery_recorder.dart (hunk 2 above)
Comment (pennzht) [reply lib/benchmarks/gallery_recorder.dart:31]: Done.

File: pubspec.yaml (hunk 3 above)
Comment (pennzht) [reply pubspec.yaml:35]: `web_benchmarks` is used by `lib/benchmarks/` as well, so it needs to be a regular dependency. [Reference](https://dart.dev/tools/pub/dependencies#dev-dependencies)

File: test/benchmarks_test.dart (hunk 4 above)
Comment (pennzht) [reply test/benchmarks_test.dart:10]: Thank you! I borrowed this test from the `web_benchmarks` test itself.

File: test/benchmarks_test.dart (hunk 5 above)
Comment (pennzht) [reply test/benchmarks_test.dart:21]: This test does not need to display the result, but `run_benchmarks.dart` does.

File: test/benchmarks_test.dart (hunk 6 above)
Comment (pennzht) [reply test/benchmarks_test.dart:4]: Reordered alphabetically. Is this correct?

**Comment by pennzht** [comment]
@guidezpl 
> Are there plans to add these benchmarks to CI in a later PR?

The next step is to add them back to `flutter/flutter`'s devicelab.

There are not yet plans to add to the CI in this repo, but it's a good idea.

File: lib/benchmarks/gallery_recorder.dart (hunk 7 above)
Comment (pennzht) [reply lib/benchmarks/gallery_recorder.dart:18]: Added documentation.

File: test/benchmarks_test.dart (hunk 4 above)
Comment (pennzht) [reply test/benchmarks_test.dart:8]: Done.

File: lib/benchmarks/gallery_recorder.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (guidezpl) [reply lib/benchmarks/gallery_recorder.dart]: Can you add this to the dartdoc? maybe `testOnlyScrolls`

File: pubspec.yaml (hunk 3 above)
Comment (guidezpl) [reply pubspec.yaml:35]: Stepping back, do the benchmarks need to be in `lib`? They're effectively tests, so if we can, we should avoid including them in builds.

File: lib/benchmarks/client.dart [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import 'package:web_benchmarks/client.dart';
+
+import 'common.dart';
+
+// Run benchmarks.
```
Comment (clocksmith) [inline lib/benchmarks/client.dart:5]: nit: this comment currently does not provide any new information over the code. Describe why/how the benchmarks are run in the main method of this program, and maybe how this program gets executed

File: test/benchmarks_test.dart [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        'apply_frame',
+        'drawFrameDuration',
+      ]) {
+        for (final valueName in <String>[
+          'average',
+          'outlierAverage',
+          'outlierRatio',
+          'noise',
+        ]) {
+          expect(
```
Comment (clocksmith) [inline test/benchmarks_test.dart:39]: This wont test against other metric + value names that might exist. If that needs to be tested, create another expect for `taskResult.scores.size/length, equals(3 * 4)` (or if the metricName and valueName lists are extracted so you are able to programmatically grab the lengths to multiply)

**Review by clocksmith — APPROVED** [review]

File: lib/benchmarks/client.dart (hunk 9 above)
Comment (pennzht) [reply lib/benchmarks/client.dart]: Thank you! I expanded the documentation.

File: test/benchmarks_test.dart (hunk 10 above)
Comment (pennzht) [reply test/benchmarks_test.dart:44]: Done. Thank you!

File: pubspec.yaml (hunk 3 above)
Comment (pennzht) [reply pubspec.yaml:35]: This is a good idea. I'm testing it out.

Comment (pennzht) [reply pubspec.yaml:35]: Done. It's looking great!

**Review by yjbanov — APPROVED** [review]
The usage of `package:web_benchmarks` LGTM. @clocksmith should review the contents of the benchmarks.

Is the plan to run these as part of Github actions? Curious about your experience.

lgtm

**Comment by pennzht** [comment]
@yjbanov 

Thank you!

I don't have any experience with Github actions yet.

> The usage of `package:web_benchmarks` LGTM. @clocksmith should review the contents of the benchmarks.
> 
> Is the plan to run these as part of Github actions? Curious about your experience.
> 
> lgtm

**Pull request merged**